### ISSUE TITLE
small range correction not working correctly

### ISSUE THREAD

**Issue by ascv** [ascv/HyperLogLog#17]
The small range correction is returning estimates that are abnormally off for a linear count.

**Comment by B3AU** [comment]
Looking forward to this!

As a temporary workaround I'm using svpcom's python implementation to do the final estimation:

```python
import HLL
from random import choice

#Sample 5000 numbers from [0..1000], i.e. ~20% uniques 
possible = range(1000)
sample = [choice(possible) for _ in range(5000)] #sample with repetition

print ("True cardinality: {}".format(len(set(sample))))

hll = HLL.HyperLogLog(14)
for s in sample:
  hll.add(str(s))
print ("Estimated cardinality (C): {}".format(hll.cardinality()))

import hyperloglog #svpcom implementation
def cardinality(hll_c):
  hll_python = hyperloglog.HyperLogLog(0.01)#uses error_rate instead of p 14 = error_rate 0.01
  #p = int(math.ceil(math.log((1.04 / error_rate) ** 2, 2))) should inverse this to get error_rate from p, and p from hll_c.size
  hll_python.M = struct.unpack_from("B"*hll_c.size(),hll_c.registers())
  return hll_python.card()

print ("Estimated cardinality (Python): {}".format(cardinality(hll)))
```
```
True cardinality: 991
Estimated cardinality (C): 1436.4178401346198
Estimated cardinality (Python): 995.648975995318
```

This gives me fast merging and adding from your implementation with the more accurate HLL++ bias correction.

**Comment by Abzac** [comment]
Any chances for solving this issue?

We use your library in our project with cardinality function fix by B3AU (using svpcom library for cardinality estimating). But this solution is pretty slow for our needs (cause svpcom's class instantition and struct unpacking is too slow).

The problem with your library is when you have small cardinalities (like less about 35k, in real life 50k), the bias is too big, like around 44-45%. I'm trying to rewrite your estimate function, just like it has to be in HLL++ Google's paper and svpcom's library but it's all too complicated for me, so I've got nothing as for now.

Do we need to look forward for some another solution or should wait for it to be fixed? Although, there are very small number of HLL libraries (for all the languages) in the Internet.

Thank you so much for your work!!

**Comment by ascv** [comment]
> The problem is when you have small cardinalities (like less than about 35k, in real life 50k), the bias is too big, like around 44-45%.

Hmmm. When you are having this problem, can you do:

    print map(int, my_hll.registers())

and paste the results in this thread (or attach a file of the output) along with the expected (true) cardinality?

**Comment by Abzac** [comment]
Sure!

The true cardinality of the set is 1649.
(1) Svpcom's HLL gives 1650, so, error_rate is 0.0606...
(2) Your HLL gives 2392, error rate is: 45.0576...
(3) Your HLL with SVPCom's cardinality estimation gives 1658, error rate is just 0.5458...

Please note that svpcom's hll uses sha1 hashing instead of murmur32, so the results (1) and (2) are never the same, but both still correct giving more or less error comparing with the other.

Here is the file attached
[ascv_hll_registers1.txt](https://github.com/ascv/HyperLogLog/files/271675/ascv_hll_registers1.txt)

**Comment by ascv** [comment]
Thanks for posting the debugging info. It was very helpful.

When you have a large number of zeros, the algorithm reverts back to doing a simple linear count. This is what was happening in your case. When I tried verifying the linear count I discovered a bug. The fix is on the v1.1 branch. Let me know if you are still having problems on that branch.

**Comment by Abzac** [comment]
Thank you so much again!

This fix really works! Would you update your lib on pypi?

**Comment by ascv** [comment]
I updated the library on pypi so it should be available through pip now. Thanks for testing and posting your results.

I edited this issue to reflect the bug in the small range correction and I am closing. I don't want users conflating the bug in the small range correction with bias correction. B3AU's work around should no longer be necessary.

**Comment by B3AU** [comment]
svpcom's implementation does some extra correction to get a smoother transition from linear counting to hll. 

Here's a plot for k=12
![svpcom vs ascv HLL](https://cloud.githubusercontent.com/assets/4571264/15581387/60261934-236c-11e6-9f01-a0f0e78ab017.jpeg)

**Comment by ascv** [comment]
I edited this issue to be about the small range correction bug. I did this b/c the workaround posted, I think intended to show the spike in the error since that was what this issue was originally about, ironically showed the bug Azbac encountered. I didn't realize this until I was looking into Azbac's bug. B/c the two issues were seeming to get conflated I edited the issue title and original post to be about the bug specifically.

Anyway, I've opened a new issue for the spike in the error, again :) See [issue 18](https://github.com/ascv/HyperLogLog/issues/21).
